### PR TITLE
Avoid casting REAL into DOUBLE on is_nan()

### PR DESCRIPTION
**Pull request by assaf2**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core query engine

> How would you describe this change to a non-technical end user or system administrator?

Instead of translating `is_nan(real_column)` into `is_nan(cast(real_column as DOUBLE))` use `is_nan(real_column)` directly, since REAL type can test for NaN values directly.

## Documentation

( ) No documentation is needed.
(X) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(X) No release notes entries required.
( ) Release notes entries required with the following suggested text:

<!--```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```-->

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 1]
```diff
@@ -1176,6 +1176,15 @@ public static boolean isNaN(@SqlType(StandardTypes.DOUBLE) double num)
         return Double.isNaN(num);
     }
 
+    @Description("Test if value is not-a-number")
+    @ScalarFunction("is_nan")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean isNaNReal(@SqlType(StandardTypes.REAL) long value)
+    {
+        float floatValue = intBitsToFloat((int) value);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1184]: i'd use `toIntExact` here.

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java (hunk 1 above)
Comment (assaf2) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1184]: I thought about it, but seems like this is how `isNaN` usually tested (at this class, at `io.trino.type.RealOperators` and at `io.trino.spi.type.RealType`). I can change it in all places (on a separated PR), WDYT?

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1184]: Yes, we're not consistent. Sometimes we do unchecked cast (`(int)`) and sometimes we do a checked one (`toIntExact`).  In new code, please use `toIntExact` unless there is a compelling reason not to.

Comment (assaf2) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java]: done

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 2]
```diff
@@ -1176,6 +1176,20 @@ public static boolean isNaN(@SqlType(StandardTypes.DOUBLE) double num)
...
+    @ScalarFunction("is_nan")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean isNaNReal(@SqlType(StandardTypes.REAL) long value)
+    {
+        try {
+            float floatValue = intBitsToFloat(toIntExact(value));
+            return Float.isNaN(floatValue);
+        }
+        catch (IllegalArgumentException e) {
+            throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE, "Out of range for integer: " + value, e);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1189]: No need to produce TrinoException for what would be an internal bug (or a bug in a connector)
(just like some ArrayIndexOutOfBound, or NPE).

Comment (assaf2) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1184]: Removed

**Pull request merged**